File: leetcode/39-combination-sum.py
# https://leetcode.com/problems/combination-sum/

from typing import List

class Solution:
	def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
		combinations = list()

		def backtracking(index, _sum, combination):
			if _sum > target:
				return
			elif _sum == target:
				combinations.append(combination.copy())
				return

			for i in range(index, len(candidates)):
				combination.append(candidates[i])
				backtracking(i, _sum + candidates[i], combination)
				combination.pop()

		backtracking(0, list(), 0)
		return combinations

File: leetcode/39-combination-sum.js
// https://leetcode.com/problems/combination-sum/

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
  const combinations = [];
  const levels = {}; // tree levels
  function backtracking(index, sum, combination, level) {
    if (!(level in levels)) levels[level] = [];
    levels[level].push([...combination]);
    if (sum > target) return;
    if (sum === target) {
      combinations.push([...combination]);
      return;
    }

    for (let i = index; i < candidates.length; i++) {
      combination.push(candidates[i]);
      backtracking(i, sum + candidates[i], combination, level+1);
      combination.pop();
    }

  }

  backtracking(0, 0, [], 0);
  console.log(levels);
  return combinations;
};



console.log(combinationSum([7, 6, 3, 2], 7));